web/src/features/datasets/server/service.ts

```diff
@@ -0,0 +1,332 @@
+  const clickhouseSession = v4();
+  try {
+    const runs = await getDatasetRunsFromPostgres(input);
+    console.log(runs);
+    await createTempTableInClickhouse(tableName, clickhouseSession);
+    await insertPostgresDatasetRunsIntoClickhouse(
+      runs,
+      tableName,
+      input.projectId,
+      input.datasetId,
+      clickhouseSession,
+    );
+
+    // these calls need to happen sequentially as there can be only one active session with
+    // the same session_id at the time.
+    const scores = await getScoresFromTempTable(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+    const obsAgg = await getObservationLatencyAndCostForDataset(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+    const traceAgg = await getTraceLatencyAndCostForDataset(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+
+    console.log("scores", scores);
+    console.log("obsAgg", obsAgg);
+    console.log("traceAgg", traceAgg);
+    await deleteTempTableInClickhouse(tableName);
+
+    const enrichedRuns = runs.map(({ run_items, ...run }) => {
+      const observation = obsAgg.find((o) => o.runId === run.run_id);
+      const trace = traceAgg.find((t) => t.runId === run.run_id);
+      return {
+        ...run,
+        projectId: input.projectId,
+        datasetId: input.datasetId,
+        id: run.run_id,
+        avgTotalCost: trace?.cost
+          ? new Decimal(trace.cost)
+          : observation?.cost
+            ? new Decimal(observation.cost)
+            : new Decimal(0),
+        countRunItems: run_items.length,
+        name: run.run_name,
+        description: run.run_description,
+        metadata: run.run_metadata,
+        createdAt: run.run_created_at,
+        updatedAt: run.run_updated_at,
+        avgLatency: trace?.latency ?? observation?.latency ?? 0,
+        scores: aggregateScores(scores.filter((s) => s.run_id === run.run_id)),
+      };
+    });
+
+    return enrichedRuns;
+  } catch (e) {
+    logger.error("Failed to fetch dataset runs from clickhouse", e);
+    await deleteTempTableInClickhouse(tableName);
+    throw e;
+  }
+};
+
+export const insertPostgresDatasetRunsIntoClickhouse = async (
+  runs: PostgresDatasetRun[],
+  tableName: string,
+  projectId: string,
+  datasetId: string,
+  clickhouseSession: string,
+) => {
+  const rows = runs.flatMap((run) =>
+    run.run_items.map((item) => ({
+      project_id: projectId,
+      run_id: run.run_id,
+      run_item_id: item.ri_id,
+      dataset_id: datasetId,
+      trace_id: item.trace_id,
+      observation_id: item.observation_id,
+    })),
+  );
+
+  await clickhouseClient({ session_id: clickhouseSession }).insert({
+    table: tableName,
+    values: rows,
+    format: "JSONEachRow",
+  });
+};
+
+export const createTempTableInClickhouse = async (
+  tableName: string,
+  clickhouseSession: string,
+) => {
+  const query = `
+      CREATE TEMPORARY TABLE IF NOT EXISTS ${tableName}
```

Comment:
we cannot use a param here. Hence, table name must not be editable by our users.